// ng-wizard/esm5/lib/core/ng-wizard.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Injectable, Optional, Inject } from '@angular/core';
import { DEFAULT_CONFIG } from '../utils/constants';
import { NG_WIZARD_CONFIG_TOKEN } from './ng-wizard-config.token';
import { Subject } from 'rxjs';
import { merge } from '../utils/functions';
import * as i0 from "@angular/core";
import * as i1 from "./ng-wizard-config.token";
import * as ɵngcc0 from '@angular/core';
var NgWizardService = /** @class */ (function () {
    function NgWizardService(config) {
        this.config = config;
        this.defaultConfig = tslib_1.__assign({}, DEFAULT_CONFIG);
        if (this.config) {
            this.defaultConfig = merge(this.defaultConfig, this.config);
        }
        // Observable sources
        this.resetWizard = new Subject();
        this.showNextStep = new Subject();
        this.showPreviousStep = new Subject();
        this.setTheme = new Subject();
        // Observable streams
        this.resetWizard$ = this.resetWizard.asObservable();
        this.showNextStep$ = this.showNextStep.asObservable();
        this.showPreviousStep$ = this.showPreviousStep.asObservable();
        this.setTheme$ = this.setTheme.asObservable();
    }
    /**
     * @return {?}
     */
    NgWizardService.prototype.getDefaultConfig = /**
     * @return {?}
     */
    function () {
        return tslib_1.__assign({}, this.defaultConfig);
    };
    /**
     * @return {?}
     */
    NgWizardService.prototype.reset = /**
     * @return {?}
     */
    function () {
        this.resetWizard.next();
    };
    /**
     * @return {?}
     */
    NgWizardService.prototype.next = /**
     * @return {?}
     */
    function () {
        this.showNextStep.next();
    };
    /**
     * @return {?}
     */
    NgWizardService.prototype.previous = /**
     * @return {?}
     */
    function () {
        this.showPreviousStep.next();
    };
    /**
     * @param {?} theme
     * @return {?}
     */
    NgWizardService.prototype.theme = /**
     * @param {?} theme
     * @return {?}
     */
    function (theme) {
        this.setTheme.next(theme);
    };
    /** @nocollapse */
    NgWizardService.ctorParameters = function () { return [
        { type: undefined, decorators: [{ type: Optional }, { type: Inject, args: [NG_WIZARD_CONFIG_TOKEN,] }] }
    ]; };
    /** @nocollapse */ NgWizardService.ngInjectableDef = i0.ɵɵdefineInjectable({ factory: function NgWizardService_Factory() { return new NgWizardService(i0.ɵɵinject(i1.NG_WIZARD_CONFIG_TOKEN, 8)); }, token: NgWizardService, providedIn: "root" });
NgWizardService.ɵfac = function NgWizardService_Factory(t) { return new (t || NgWizardService)(ɵngcc0.ɵɵinject(NG_WIZARD_CONFIG_TOKEN, 8)); };
NgWizardService.ɵprov = ɵngcc0.ɵɵdefineInjectable({ token: NgWizardService, factory: function (t) { return NgWizardService.ɵfac(t); }, providedIn: 'root' });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(NgWizardService, [{
        type: Injectable,
        args: [{
                providedIn: 'root'
            }]
    }], function () { return [{ type: undefined, decorators: [{
                type: Optional
            }, {
                type: Inject,
                args: [NG_WIZARD_CONFIG_TOKEN]
            }] }]; }, null); })();
    return NgWizardService;
}());
export { NgWizardService };
if (false) {
    /** @type {?} */
    NgWizardService.prototype.resetWizard$;
    /** @type {?} */
    NgWizardService.prototype.showNextStep$;
    /** @type {?} */
    NgWizardService.prototype.showPreviousStep$;
    /** @type {?} */
    NgWizardService.prototype.setTheme$;
    /**
     * @type {?}
     * @private
     */
    NgWizardService.prototype.resetWizard;
    /**
     * @type {?}
     * @private
     */
    NgWizardService.prototype.showNextStep;
    /**
     * @type {?}
     * @private
     */
    NgWizardService.prototype.showPreviousStep;
    /**
     * @type {?}
     * @private
     */
    NgWizardService.prototype.setTheme;
    /**
     * @type {?}
     * @private
     */
    NgWizardService.prototype.defaultConfig;
    /**
     * @type {?}
     * @private
     */
    NgWizardService.prototype.config;
}
